### PR TITLE
Change the selection of downgrade version in TestUpgradeBrokenPackageVersion

### LINKED ISSUE #3454: Flaky test: TestUpgradeBrokenPackageVersion
## Flaky Test

* **Test Name:** TestUpgradeBrokenPackageVersion

* **Link:** https://github.com/elastic/elastic-agent/blob/b1d2e6b04062b8572718e583590afe678579bf9d/testing/integration/upgrade_test.go#L868

* **Branch:** https://github.com/elastic/elastic-agent/tree/main
* **Artifact Link:** https://buildkite.com/elastic/elastic-agent/builds/3422#018ab6de-5d5c-4de6-949e-0f962b039a59
* **Notes:** 
The test tries an upgrade to itself depending on the output of the artifact API (it just takes the penultimate version)

### Stack Trace

```
paste stack trace here
```

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?
Fixes selection of the elastic agent version to downgrade to in the test so that we don't try to upgrade the agent on its own version

## Why is it important?
The test outcome today depends on the order of versions returned by Artifact API making it flaky when there are some changes.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Author's Checklist

- [ ]

## Related issues

- Closes #3454 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: testing/integration/upgrade_test.go [hunk 1]
```diff
@@ -936,18 +940,22 @@ func TestUpgradeBrokenPackageVersion(t *testing.T) {
 	versionList, err := aac.GetVersions(ctx)
 	require.NoError(t, err)
 	require.NotEmpty(t, versionList.Versions, "Artifact API returned no versions")
-	latestVersion := versionList.Versions[len(versionList.Versions)-1]
 
-	t.Logf("Upgrading to version %q", latestVersion)
+	// transform and reverse the version list and find the most recent version that is different from the broken version returned
+
+	// look for 1 version with the current major that is lower than the current one
```
Comment (AndersonQ) [inline testing/integration/upgrade_test.go:946]: I'm guessing it's 1 thing? AM I right?

```suggestion
	// transform and reverse the version list and find the most recent version that is different from the broken version returned
	// look for 1 version with the current major that is lower than the current one
```

**Review by AndersonQ — APPROVED** [review]

**Review by blakerouse — CHANGES_REQUESTED** [review]
This test overall has a problem. Being that this test is downgrading the installed build to a snapshot build, means that when 'elastic-agent uninstall' is ran at the end of the test will be the snapshots uninstall. 

The snapshot uninstall might not have the required code to kill the watcher, which then means the watcher is still running and will break a following test.

**Comment by pchila** [comment]
@blakerouse since we start from the latest snapshot for test, we don't have that many choices about versions to move to.
The purpose of the test is to ensure that an agent install that has a broken/missing package version file can still be upgraded as [requested when the feature was implemented](https://github.com/elastic/elastic-agent/pull/2908#discussion_r1237153021)
If we cannot downgrade here, what do you suggest? Remove the test entirely? Skip it for a few release cycles?
If we don't merge this PR we still have the test just in the flaky version which is probably a worse outcome

**Comment by blakerouse** [comment]
> @blakerouse since we start from the latest snapshot for test, we don't have that many choices about versions to move to. The purpose of the test is to ensure that an agent install that has a broken/missing package version file can still be upgraded as [requested when the feature was implemented](https://github.com/elastic/elastic-agent/pull/2908#discussion_r1237153021) If we cannot downgrade here, what do you suggest? Remove the test entirely? Skip it for a few release cycles? If we don't merge this PR we still have the test just in the flaky version which is probably a worse outcome

@pchila We have an overall problem with all the upgrade tests in the integration suite. They are either not testing the actual agent under test (https://github.com/elastic/elastic-agent/issues/3461) or the ending Elastic Agent is an older version of Elastic Agent that might not have required fixes for the testing suite to pass (in this case missing watcher kill).

> @blakerouse would running it in isolation address your concerns? for the time being. once we have previous major with proper watcher capabilities we would go back to normal runs

@michalpristas I wish that would work, but uninstall is a core part of the testing framework. Running this test in isolation will still result in it failing on Windows, because it will not be able to kill the watcher.

Only possible solution is to wait until the watcher is completely done running, before performing the uninstall. I think that is the only valid solution. This is the only case that works, that is still assuming that we don't hit an issue with `RemovePath` on the older version of the Elastic Agent.

**Comment by cmacknz** [comment]
> @pchila We have an overall problem with all the upgrade tests in the integration suite. They are either not testing the actual agent under test (https://github.com/elastic/elastic-agent/issues/3461) or the ending Elastic Agent is an older version of Elastic Agent that might not have required fixes for the testing suite to pass (in this case missing watcher kill).

Fundamentally to solve this we need the ability to upgrade to the version of the agent that was built in the PR. We need https://github.com/elastic/elastic-agent/issues/2780.

We need to move tests between previously released versions and official snapshots onto a schedule because the code in PRs doesn't affect them at all.


> Only possible solution is to wait until the watcher is completely done running, before performing the uninstall. I think that is the only valid solution. This is the only case that works, that is still assuming that we don't hit an issue with RemovePath on the older version of the Elastic Agent.

Agreed I think this is the only way to deal with this right now, besides skipping the test or figuring out how to unit test this instead of integration testing it.

**Comment by pchila** [comment]
> Fundamentally to solve this we need the ability to upgrade to the version of the agent that was built in the PR. We need https://github.com/elastic/elastic-agent/issues/2780.

For standalone we mostly do, except for versions of agent where we cannot specify a `file://` source-uri and a `skip-verify` flag. For Fleet managed tests I would expect the same but I didn't actually verify that we do use the packaged version of agent something can be improved there.

For the scope of this PR, do we want to include all of those changes? Or do we create a separate issues/PRs for those ?

**Comment by pchila** [comment]
@blakerouse @cmacknz Adding the wait means partially revert commit 7e86d24cd336145b9641f621a3de10ed3af0c9fe where the cleanup functions have been removed.
Would that be acceptable for the moment in this PR ?

**Comment by pchila** [comment]
@blakerouse could you give another review pass now that we have reintroduced the wait for watcher on the upgrade tests ?

**Comment by pchila** [comment]
@cmacknz 

> We need to move tests between previously released versions and official snapshots onto a schedule because the code in PRs doesn't affect them at all.

https://github.com/elastic/elastic-agent/issues/3475

**Comment by blakerouse** [comment]
@pchila See my PR #3477

I believe I have upgrade testing working correctly, even this one.

**Comment by pchila** [comment]
This is replaced by https://github.com/elastic/elastic-agent/pull/3477

**Pull request closed without merging**